subworkflows/local/utils_nfcore_mcmicro_pipeline/main.nf

```diff
@@ -181,60 +185,166 @@ def validateInputParameters() {
-        ctr = ctr + 1
+// function that returns the index of a given column from a given sheet
+//   as defined in the schema file.
+//   (will need to be updated when the schema files change)
+def input_sheet_index( sheet_type, column_name ) {
+
+    def index_map = [:]
+    if (sheet_type == "sample") {
+        index_map = [sample: 0, image_directory: 1, cycle_images: 2, dfp: 3, ffp: 4]
+    } else if (sheet_type == "cycle") {
+        index_map = [sample: 0, cycle_number: 1, channel_count: 2, image_tiles: 3, dfp: 4, ffp: 5]
+    } else if (sheet_type == "marker") {
+        index_map = [channel_number: 0, cycle_number: 1, marker_name: 2, filter: 3,
+                        excitation_wavelength: 4, emission_wavelength: 5]
+    } else {
+        error("Error: bad sheet type: $sheet_type")
     }
-    return sample_sheet_index_map
+
+    return index_map[column_name]
 }
 
-def make_ashlar_input_sample(ArrayList sample_sheet_row, Map sample_sheet_index_map) {
-    sample_name_index = sample_sheet_index_map['sample']
-    image_dir_path_index = sample_sheet_index_map['image_directory']
-    if (sample_sheet_index_map.keySet().collect().contains("cycle_images")) {
-        tmp_path = sample_sheet_row[image_dir_path_index]
+def validateInputSamplesheetRow ( row, mode ) {
+    // TODO: Add sample sheet validation for cycle inputs
+
+    if (mode == "sample") {
+        // check for the existence of all files under cycle_image column in the given image_directory
+        if (row.size() >= 3 && row[2] != []) {
+            def file_list = row[2].split(" ")
+            file_list.each { curr_file ->
+                def curr_path = new File(row[1].toString() + "/" + curr_file)
+                if (!curr_path.exists()) {
+                    error("Error: file in samplesheet not found: $curr_path")
+                }
+            }
+        }
+    }
+
+    return row
+}
+
+def validateInputSamplesheetMarkersheet ( sheet_data, mode ) {
+    if (mode == 'cycle' ) {
+        def ctr = 0
+        def sample_cycle_list = []
+        def marker_cycle_list = []
+        def idx_sample_cycle = input_sheet_index("cycle", "cycle_number")
+        def idx_marker_cycle = input_sheet_index("marker", "cycle_number")
+
+        sheet_data.each { curr_list ->
+            if (ctr == 0) {
+                curr_list.each { curr_sublist ->
+                    sample_cycle_list.add(curr_sublist[idx_sample_cycle])
+                }
+            } else {
+                marker_cycle_list.add(curr_list[idx_marker_cycle])
+            }
+            ctr++
+        }
+        if (marker_cycle_list.unique() != sample_cycle_list.unique() ) {
+            error("Error: cycle_number in sample and marker sheets must match 1:1!")
+        }
+    } else if ( mode == 'sample' ) {
+        // TODO: add validation for 1 row per sample samplesheet & markersheet correspondence
+    } else {
+        error("Error: bad mode $mode in validateInputSamplesheetMarkersheet()")
```

Comment:
I'm not sure we need to guard against this given that the schema has already validated we've got one of the allowed values. Regardless, we don't need the function name since the stack trace already contains that. Also we don't need `Error:` at the beginning of the error message (see many other instances of this as well).

Reply:
Agreed about the `Error:` strings. I've removed those

Maybe we should just remove the `mode` parameter from both `validateInputSampesheetRow` and `validateInputSamplesheetMarkersheet`.  We only use the former for 1 row per sample samplesheets and we only use the latter for 1 row per sample per cycle samplesheets.
How does this sound?

Reply:
This only applies to `validateInputSamplesheetMarkersheet` now. I removed `validateInputSampesheetRow` because it was only used as part of the `cycle_images` concept.

Reply:
Agreed. I've removed this.
The other question I ask is discussed in another conversion, so marking this one as resolved.